testing/tools/integration/observability.py

```diff
@@ -245,3 +291,116 @@ def run(self):
+                stdout = r.read()
+                if not stdout:
+                    time.sleep(0.1)
+                    continue
+                else:
+                    if self.pattern.search(stdout):
+                        logging.debug('Application reports it is ready.')
+                        self.stop()
+                        break
+                if time.time() - started > self.timeout:
+                    outputs = runners_output_format(self.runners)
+                    self.error = TimeoutError(
+                        'Application did not report as ready after {} '
+                        'seconds. It had the following outputs:\n===\n{}'
+                        .format(self.timeout, outputs))
+                    self.stop()
+                    break
+
+
+class RunnerChecker(StoppableThread):
+    __base_name__ = 'RunnerChecker'
+
+    def __init__(self, runner, patterns, timeout=30, start_from=0):
+        super(RunnerChecker, self).__init__()
+        self.name = self.__base_name__
+        self.runner = runner
+        self.runner_name = runner.name
+        self.start_from = start_from
+        self._path = runner.file.name
+        self.timeout = timeout
+        self.error = None
+        if isinstance(patterns, (list, tuple)):
+            self.patterns = patterns
+        else:
+            self.patterns = [pattern]
+        self.compiled = [re.compile(p) for p in patterns]
+
+    def run(self):
+        with open(self._path, 'rb') as r:
+            last_match = self.start_from
+            started = time.time()
+            while not self.stopped():
+                r.seek(last_match)
+                stdout = r.read()
+                if not stdout:
+                    time.sleep(0.1)
+                    continue
+                else:
+                    match = self.compiled[0].search(stdout)
+                    if match:
+                        logging.debug('Pattern %r found in runner STDOUT.'
+                                      % match.re.pattern)
+                        self.compiled.pop(0)
+                        last_match = self.start_from
+                        if self.compiled:
+                            continue
+                        self.stop()
+                        break
+                if time.time() - started > self.timeout:
+                    r.seek(self.start_from)
+                    stdout = r.read()
+                    self.error = TimeoutError(
+                        'Runner {!r} did not have patterns {!r}'
+                        ' after {} seconds.'
+                        .format(self.runner_name,
+                                [rx.pattern for rx in self.compiled],
+                                self.timeout))
+                    self.stop()
+                    break
+
+
+####################################
+# Query response parsing functions #
+####################################
+def joined_partition_query_data(responses):
```

Comment:
`join` appears again in another context. Maybe there's a better term here? `coalesce` ?
Also `data` is unclear:
`coalesce_partition_query_responses` ?

Reply:
`merge_paritition_query_responses` ?